lls/cache.c

```diff
@@ -0,0 +1,428 @@
+	lls_entry->ts = time(NULL);
+
+send:
+	req->entry = lls_entry;
+	req->op = op;
+	req->cache = cache;
+	memcpy(req->ip_key, ip_key, cache->key_len);
+
+	ret = mb_send_entry(mb, req);
+	if (ret != 0 && op != LLS_CACHE_DEC)
+		rte_mempool_put(cache->entry_pool, lls_entry);
+	return ret;
+}
+
+void
+lls_cache_scan(struct lls_cache *cache, struct gatekeeper_if *front,
+	struct gatekeeper_if *back)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	const void *ip_key;
+	void *data;
+	time_t now = time(NULL);
+
+	index = rte_hash_iterate(cache->hash, &ip_key, &data, &iter);
+	while (index >= 0) {
+		struct lls_cache_entry *lls_entry = cache->data[index];
+		uint32_t timeout;
+
+		/*
+		 * If an entry is already stale or invalid, continue
+		 * to try to resolve it while there's interest.
+		 */
+		if (!lls_entry->valid || lls_entry->stale) {
+			if (lls_entry->refcnt > 0) {
+				lls_entry->valid = false;
+				/* Broadcast request on both ports. */
+				cache->xmit_req(front->id, ip_key, NULL);
+				cache->xmit_req(back->id, ip_key, NULL);
+			} else
+				__lls_cache_del(cache, ip_key);
+			goto next;
+		}
+
+		if (lls_entry->port_id == front->id)
+			timeout = cache->front_timeout_sec;
+		else if (back != NULL && lls_entry->port_id == back->id)
+			timeout = cache->back_timeout_sec;
+		else {
+			RTE_LOG(ERR, GATEKEEPER, "lls: %s cache has an entry associated with neither the front nor back port (%hhu)\n",
+				cache->name, lls_entry->port_id);
+			goto next;
+		}
+
+		if (now - lls_entry->ts > timeout) {
+			lls_entry->stale = true;
+			if (lls_entry->refcnt > 0) {
+				/* Unicast request. */
+				cache->xmit_req(lls_entry->port_id, ip_key,
+					&lls_entry->ha);
+			}
+		}
+next:
+		index = rte_hash_iterate(cache->hash, &ip_key, &data, &iter);
+	}
+
+	if (get_lls_conf()->debug)
+		lls_cache_dump(cache);
+}
+
+void
+lls_cache_destroy(struct lls_cache *cache)
+{
+	rte_mempool_free(cache->entry_pool);
+	rte_hash_free(cache->hash);
+	rte_free(cache->data);
+}
+
+int
+lls_cache_init(struct lls_config *lls_conf, struct lls_cache *cache)
+{
+	unsigned int socket_id = rte_lcore_to_socket_id(lls_conf->lcore_id);
+	struct rte_hash_parameters lls_cache_params = {
+		.name = cache->name,
+		.entries = LLS_CACHE_ENTRIES,
+		.reserved = 0,
+		.key_len = cache->key_len,
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+		.socket_id = socket_id,
+		.extra_flag = 0,
+	};
+	char entry_pool_name[128];
+
+	RTE_ASSERT(cache->key_len <= LLS_CACHE_MAX_KEY_LEN);
+
+	cache->data = rte_calloc(NULL, LLS_CACHE_ENTRIES,
+		sizeof(*cache->data), 0);
+	if (cache->data == NULL) {
+		RTE_LOG(ERR, MALLOC, "Could not allocate %s cache data\n",
+			cache->name);
+		return -1;
+	}
+
+	cache->hash = rte_hash_create(&lls_cache_params);
+	if (cache->hash == NULL) {
+		RTE_LOG(ERR, HASH, "Could not create %s cache hash\n",
+			cache->name);
+		goto data;
+	}
+
+	RTE_ASSERT(snprintf(entry_pool_name, sizeof(entry_pool_name),
+		"%s_entry_pool", cache->name < sizeof(entry_pool_name)));
```

Comment:
This code is wrong. It's funny it compiles. Look for an example in the source of Gatekeeper.